File: src/main/java/com/gitlab/lae/intellij/actions/tree/json/ActionNodeParser.java
package com.gitlab.lae.intellij.actions.tree.json;

import com.gitlab.lae.intellij.actions.tree.ActionNode;
import com.gitlab.lae.intellij.actions.tree.When;
import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

import javax.swing.*;
import java.io.IOException;
import java.io.Reader;
import java.nio.file.Path;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.IntSupplier;

import static com.gitlab.lae.intellij.actions.tree.json.JsonObjects.*;
import static java.nio.file.Files.newBufferedReader;
import static java.util.stream.StreamSupport.stream;

public final class ActionNodeParser {

    private static final String WHEN = "when";
    private static final String WHEN_ANY = "any";
    private static final String WHEN_ALL = "all";

    private ActionNodeParser() {
    }

    private static final Gson gson = new Gson();

    public static List<ActionNode> parseJsonActions(Path path)
            throws IOException {
        try (Reader reader = newBufferedReader(path)) {
            return parseJsonActions(reader);
        }
    }

    public static List<ActionNode> parseJsonActions(Reader reader) {
        JsonElement element = gson.fromJson(reader, JsonElement.class);
        ActionNode action = toActionNode(
                element,
                new AtomicInteger()::getAndIncrement
        );
        return action.items();
    }

    private static ActionNode toActionNode(
            JsonElement element,
            IntSupplier seq
    ) {
        JsonObject o = element.getAsJsonObject();
        String id = removeString(o, "id", () -> "ActionsTree" + seq.getAsInt());
        String sep = removeString(o, "separator-above", () -> null);
        String name = removeString(o, "name", () -> "Unnamed");
        boolean sticky = removeBoolean(o, "sticky", () -> false);
        When when = processWhen(o.remove(WHEN));

        List<KeyStroke> keys =
                removeArray(o, "keys", ActionNodeParser::toKeyStroke);

        List<ActionNode> items =
                removeArray(o, "items", it -> toActionNode(it, seq));

        if (!o.keySet().isEmpty()) {
            throw new IllegalArgumentException(
                    "Invalid elements: " + o.keySet());
        }
        return ActionNode.create(id, name, sep, sticky, when, keys, items);
    }

    private static When processWhen(JsonElement element) {
        if (element == null) {
            return When.ALWAYS;
        }

        if (element.isJsonPrimitive()) {
            return When.parse(element.getAsString());
        }

        JsonObject object = element.getAsJsonObject();
        if (object.keySet().isEmpty()) {
            throw new IllegalArgumentException(
                    "'" + WHEN + "' object is empty");
        }

        if (object.keySet().size() != 1) {
            throw new IllegalArgumentException("" +
                    "'" + WHEN + "' object must only have either " +
                    "'" + WHEN_ANY + "' or " +
                    "'" + WHEN_ALL + "' element: " + object);
        }

        JsonElement any = object.remove(WHEN_ANY);
        if (any != null) {
            return When.any(processWhens(any.getAsJsonArray()));
        }

        JsonElement all = object.remove(WHEN_ALL);
        if (all != null) {
            return When.all(processWhens(all.getAsJsonArray()));
        }

        throw new IllegalArgumentException("" +
                "'" + WHEN + "' object must only have either " +
                "'" + WHEN_ANY + "' or " +
                "'" + WHEN_ALL + "' element: " + object);
    }

    private static When[] processWhens(JsonArray clauses) {
        return stream(clauses.getAsJsonArray().spliterator(), false)
                .map(ActionNodeParser::processWhen)
                .toArray(When[]::new);
    }

    private static KeyStroke toKeyStroke(JsonElement element) {
        KeyStroke key = KeyStroke.getKeyStroke(element.getAsString());
        if (key == null) {
            throw new IllegalArgumentException(
                    "Invalid key stroke: " + element.getAsString());
        }
        return key;
    }
}

File: src/main/kotlin/com/gitlab/lae/intellij/actions/tree/json/ActionNodeParser.kt
package com.gitlab.lae.intellij.actions.tree.json

import com.gitlab.lae.intellij.actions.tree.ActionNode
import com.gitlab.lae.intellij.actions.tree.When
import com.gitlab.lae.intellij.actions.tree.When.All
import com.gitlab.lae.intellij.actions.tree.When.Any
import com.google.gson.Gson
import com.google.gson.JsonArray
import com.google.gson.JsonElement
import java.io.Reader
import java.nio.file.Files.newBufferedReader
import java.nio.file.Path
import java.util.concurrent.atomic.AtomicInteger
import javax.swing.KeyStroke
import javax.swing.KeyStroke.getKeyStroke

private const val WHEN = "when"
private const val WHEN_ANY = "any"
private const val WHEN_ALL = "all"
private val gson = Gson()

fun parseJsonActions(path: Path): List<ActionNode> = newBufferedReader(path).use(::parseJsonActions)

fun parseJsonActions(reader: Reader): List<ActionNode> {
  val element = gson.fromJson(reader, JsonElement::class.java)
  val seq = AtomicInteger()
  val (_, _, _, _, _, _, items) = toActionNode(element, seq::getAndIncrement)
  return items
}

private fun toActionNode(element: JsonElement, seq: () -> Int): ActionNode {
  val o = element.asJsonObject
  val id = o.removeString("id") { "ActionsTree.Node." + seq() }
  val sep = o.remove("separator-above")?.asString
  val name = o.removeString("name") { "Unnamed" }
  val sticky = o.removeBoolean("sticky") { false }
  val condition = processWhen(o.remove(WHEN))
  val keys = o.removeArray("keys", ::toKeyStroke)
  val items = o.removeArray("items") { toActionNode(it, seq) }
  require(o.keySet().isEmpty()) { "Invalid elements: " + o.keySet() }
  return ActionNode(id, name, sep, sticky, condition, keys, items)
}

private fun processWhen(element: JsonElement?): When {
  element ?: return When.Always

  if (element.isJsonPrimitive) {
    return When.parse(element.asString)
  }

  val obj = element.asJsonObject
  require(obj.keySet().isNotEmpty()) { "'$WHEN' object is empty" }
  require(obj.keySet().size == 1) {
    "'$WHEN' object must only have either '$WHEN_ANY' or '$WHEN_ALL' element: $obj"
  }

  val any = obj.remove(WHEN_ANY)
  if (any != null) {
    return Any(*processWhens(any.asJsonArray))
  }

  val all = obj.remove(WHEN_ALL)
  if (all != null) {
    return All(*processWhens(all.asJsonArray))
  }

  throw IllegalArgumentException(
      "'$WHEN' object must only have either '$WHEN_ANY' or '$WHEN_ALL' element: $obj",
  )
}

private fun processWhens(clauses: JsonArray): Array<When> =
    clauses.asJsonArray.map(::processWhen).toTypedArray()

private fun toKeyStroke(element: JsonElement): KeyStroke =
    getKeyStroke(element.asString)
        ?: throw IllegalArgumentException("Invalid key stroke: " + element.asString)
